### ISSUE TITLE
[FEA] Use CUDA IPC for sharing data between Scala and Python processes.

### ISSUE THREAD

**Issue by trivialfis** [NVIDIA/spark-rapids#5561]
Hi, I'm curious whether it's possible to share GPU data between spark and pyspark using [CUDA IPC](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#interprocess-communication). Currently, the python UDF accepts [only pandas iterator](https://github.com/NVIDIA/spark-rapids/blob/aabe71a41ffdce5b00f4c1e1563222f72730aada/integration_tests/src/main/python/udf_cudf_test.py#L87). Also, based on the conversation I can find here, to transfer the data, a full serialization using arrow is required.

**Describe the solution you'd like**

I would like to skip the data serialization and data copies between GPU and CPU and between different processes. One such possibility might be using CUDA IPC to share the GPU memory. However, this handles only the low-level sharing, for higher-level functions like `MapInPandas` there might be a breaking change or rapids will have to own its unique API function for mapping cuDF data.

**Comment by revans2** [comment]
Yes, technically it is possible, but there are a lot of challenges with it. The biggest challenge is not moving the data, but in how do we partition the GPU properly so performance is not horrible. We support having python processes use the same GPU that the main Spark worker process is also using.

https://nvidia.github.io/spark-rapids/docs/additional-functionality/rapids-udfs.html#gpu-support-for-pandas-udf

But if all you want to do is a simple MapInPandas operation you have to statically partition the GPUs memory between the python processes and the Spark process. On a 16 GiB card like a v100 it can be the difference between running 2 tasks in parallel and only being able to run a single task at a time on the GPU. This has a big impact for overall performance. It can also impact how much we end up spilling to host memory or disk for out of core operations like sorting. This has a large impact on all GPUs, if they hit those cases. But if what you are primarily doing is ML/DL using python APIs it can make these use cases possible without needing a separate GPU for them in python.

> a full serialization using arrow is required

That is true, but happily CUDF operates on data that is in an arrow like format already so serializing to/from arrow is much cheaper than doing it on the CPU where it goes from Spark's UnsafeRow to arrow before being serialized with the java arrow API. We have seen some really great performance improvements by going from CUDF to Arrow on the CPU to the serialized arrow format using the native arrow library.

If you have a specific use case we are happy to work with you on it.

**Comment by trivialfis** [comment]
Thank you for the detailed information. I'm looking into the possibility of using Python interface to implement XGBoost pyspark interface, one of the concerns is that there might be overhead during data serialization, which is not just performance concern but also CPU and GPU memory usage. XGBoost has to load the data in full instead of batches, being memory hungry is one of its most significant blocker for users. As a result we hope that there's a way to reduce data copy as much as possible.

**Comment by revans2** [comment]
There is a java/scala XGBoost interface that bypasses those issues. We have been testing it in integration with the RAPIDs Accelerator and there is no data movement. It stays on the GPU and plays nicely with RAPIDs.

https://github.com/NVIDIA/spark-rapids-examples/tree/branch-22.06/examples/XGBoost-Examples

**Comment by trivialfis** [comment]
Yes. It has been a heated debate between using the Python interface and the Java interface.  There are many concerns around the JAVA interface as well, like dependencies management and user defined objective/metric/callback that has to be run on Python process, synchronization between Python XGB package and JVM package, etc.

I'm thinking that the cost of serialization might be worth it. It's out of scope for XGBoost to work with these issues. But also want to understand better if there's a way to avoid the overhead in the future as this is not a XGB specific issue and might benefit many other workflows as well.

**Comment by revans2** [comment]
You have a number of options. Sorry that this is so long and rambling, I don't have a ton of time to clean it up before sending it.

1. Disable RMM pooling and use python for XGBoost. The RAPIDs Accelerator will use a little bit of memory that XGBoost will not be able to get access to, but most of the memory should be available to it. The downside is that the RAPIDs Accelerator will be much slower because it will have to call cudaMalloc and cudfFree directly instead of using RMM pooling. The data transfer should be relatively fast, but if it is not we can work with you on doing a prototype with CUDA IPC (see option 4).
2. Statically split the memory between the Python process and the scala process. This will speed up the RAPIDs Accelerator, but a lot less memory will be available for XGBoost to use.The data transfer should still be relatively fast compared to what we have on the CPU. Again if it is not good enough we can look at doing an one off prototype.
3. Disable the RAPIDs Accelerator entirely. This will give all of the GPU memory to XGBoost, but the transfer will be much slower and there isn't much we can do about it. You also don't get any GPU acceleration for reading the data or feature engineering before training.
4. Roll your own solution as a prototype. We would read in the data/do feature engineering/etc and then hand that off to some custom code in scala that we could have python bindings for. It would get an `RDD[Table]` using [this](https://nvidia.github.io/spark-rapids/docs/additional-functionality/ml-integration.html). Then we can put in the barrier, and do a mapPartitions that would launch a custom python process, hopefully using most of the same code that Spark uses for python processes. Then it would handle doing CUDA IPC for each table over to the python process and release memory back as we go. At the end it would release any final remaining memory, tell the python process to do the training and wait for something to be sent back. That is what I have had in the back of my mind for a while on how we could do this, but I have been waiting on the async allocator and  CUDA IPC to be working properly together. From what I understand the async allocator has issues with cuda IPC that they are still working on, but it is not ready yet.

**Comment by trivialfis** [comment]
Thank you for the suggestions! Really appreciate that.

XGBoost does support drawing memory from RMM so that's not an issue. I think the last point sufficiently shows that there's a way to improve in the future and I have high hope that it will get good performance once those blockers are sorted out.
